File: server/app/schemas/imports.py
from marshmallow import EXCLUDE, Schema, fields, validate


class ColumnMappingSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    csv_column = fields.Str(required=True)
    target_field = fields.Str(required=True, description="'date', 'category', or 'ignore'")
    category_id = fields.Int(allow_none=True, description="MenuCategory.id (integer)")


class DateConfigSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    mode = fields.Str(description="'from_file', 'align_week', or 'start_date'")
    start_date = fields.Str(allow_none=True, description="Start date (YYYY-MM-DD)")
    skip_weekends = fields.Bool()
    date_format = fields.Str(allow_none=True)
    auto_detect_tags = fields.Bool()


class ImportUploadSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    file_id = fields.Str(dump_only=True)
    filename = fields.Str(dump_only=True)
    columns = fields.List(fields.Str())
    preview_rows = fields.List(fields.Dict())
    row_count = fields.Int()
    detected_delimiter = fields.Str(allow_none=True)
    auto_mapping = fields.Dict()
    detected_date_format = fields.Str(allow_none=True)


class ImportPreviewSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    file_id = fields.Str(required=True)
    column_mapping = fields.List(fields.Nested(ColumnMappingSchema))
    date_config = fields.Nested(DateConfigSchema)
    restaurant_id = fields.Int(allow_none=True)


class ImportConfirmSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    file_id = fields.Str(required=True)
    column_mapping = fields.List(fields.Nested(ColumnMappingSchema))
    date_config = fields.Nested(DateConfigSchema)
    duplicate_action = fields.Str(description="'skip', 'replace', or 'merge'")
    auto_publish = fields.Bool()
    restaurant_id = fields.Int(allow_none=True)


# ============================================================
# IMPORT CATALOGUE — liste de plats
# ============================================================

class CatalogImportUploadSchema(Schema):
    """Réponse d'upload : colonnes détectées + suggestion de colonne nom."""
    class Meta:
        unknown = EXCLUDE
    file_id = fields.Str(dump_only=True)
    filename = fields.Str(dump_only=True)
    columns = fields.List(fields.Str())
    preview_rows = fields.List(fields.Dict())
    row_count = fields.Int()
    delimiter = fields.Str(allow_none=True)
    suggested_name_column = fields.Str(allow_none=True)
    is_catalog_export = fields.Bool()
    category_paths = fields.List(fields.Str())
    known_categories = fields.Dict(keys=fields.Str(), values=fields.Int())


class CatalogImportPreviewSchema(Schema):
    """Paramètres de prévisualisation/confirmation d'un import catalogue.

    Deux modes : le fichier quelconque (name_column + category_id), et l'export
    Mariam, qui porte ses catégories et sa taxonomie et ne demande qu'une
    correspondance de catégories.
    """
    class Meta:
        unknown = EXCLUDE
    file_id = fields.Str(required=True)
    name_column = fields.Str(load_default=None, allow_none=True, description="Colonne contenant le nom du plat")
    tag_columns = fields.List(fields.Str(), load_default=[], description="Colonnes scannées pour les tags/labels")
    category_id = fields.Int(load_default=None, allow_none=True, validate=validate.Range(min=1))
    auto_detect_tags = fields.Bool(load_default=True)
    category_map = fields.Dict(
        keys=fields.Str(), values=fields.Str(), load_default=None, allow_none=True,
        description="Chemin du fichier → id de catégorie ou 'create'",
    )


class CatalogImportPreviewDishSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    name = fields.Str()
    tags = fields.List(fields.Str())
    certifications = fields.List(fields.Str())
    is_duplicate = fields.Bool()
    category_path = fields.Str()


class CatalogImportPreviewResultSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    dishes = fields.List(fields.Nested(CatalogImportPreviewDishSchema))
    total = fields.Int()
    new_count = fields.Int()
    duplicate_count = fields.Int()
    categories_to_create = fields.List(fields.Str())


# Confirmation : mêmes paramètres que la prévisualisation
CatalogImportConfirmSchema = CatalogImportPreviewSchema


class CatalogImportResultSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    created_count = fields.Int()
    skipped_count = fields.Int()

File: client/src/lib/api/imports.ts
import { api } from './client';
import type { MenuItem } from './menus';

export interface CsvUploadResponse {
  file_id: string;
  filename: string;
  columns: string[];
  preview_rows: Record<string, string>[];
  row_count: number;
  detected_delimiter: string | null;
  auto_mapping: {
    date?: string;
    categories?: Record<string, number>; // csv column → category id
  };
  detected_date_format?: string;
}

export interface ColumnMapping {
  csv_column: string;
  target_field: 'date' | 'category' | 'ignore';
  category_id?: number;
}

export interface DateConfig {
  mode: 'from_file' | 'align_week' | 'start_date';
  start_date?: string;
  skip_weekends: boolean;
  date_format?: string;
  auto_detect_tags?: boolean;
}

export interface CsvPreviewItem {
  category_id: number;
  name: string;
  order: number;
  tags: string[];
  certifications: string[];
}

export interface ImportPreviewResponse {
  menus: {
    date: string;
    date_display: string;
    items: CsvPreviewItem[];
    has_duplicate: boolean;
    existing_menu?: {
      id: number;
      status: string;
      items: MenuItem[];
    };
  }[];
  total_count: number;
  duplicates_count: number;
  new_count: number;
}

export interface ImportConfirmRequest {
  file_id: string;
  column_mapping: ColumnMapping[];
  date_config: DateConfig;
  duplicate_action: 'skip' | 'replace' | 'merge';
  auto_publish: boolean;
  restaurant_id?: number;
}

export interface ImportConfirmResponse {
  success: boolean;
  imported_count: number;
  replaced_count: number;
  skipped_count: number;
  message: string;
}

export interface CatalogImportUploadResponse {
  file_id: string;
  filename: string;
  columns: string[];
  preview_rows: Record<string, string>[];
  row_count: number;
  delimiter: string | null;
  suggested_name_column: string | null;
  /** The file carries the columns the catalogue export writes. */
  is_catalog_export: boolean;
  category_paths: string[];
  known_categories: Record<string, number>;
}

export interface CatalogImportPreviewDish {
  name: string;
  tags: string[];
  certifications: string[];
  is_duplicate: boolean;
  category_path?: string;
}

export interface CatalogImportPreviewResponse {
  dishes: CatalogImportPreviewDish[];
  total: number;
  new_count: number;
  duplicate_count: number;
  categories_to_create: string[];
}

export interface CatalogImportParams {
  file_id: string;
  name_column?: string;
  tag_columns?: string[];
  category_id?: number;
  auto_detect_tags?: boolean;
  /** Export file only: each category path of the file, to an id or 'create'. */
  category_map?: Record<string, string>;
}

export interface CatalogImportResult {
  created_count: number;
  skipped_count: number;
}

export const csvImportApi = {
  upload: async (file: File): Promise<CsvUploadResponse> => {
    const formData = new FormData();
    formData.append('file', file);
    const response = await api.post('/imports/menus/upload', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
      timeout: 30000,
    });
    return response.data;
  },

  preview: async (
    fileId: string,
    columnMapping: ColumnMapping[],
    dateConfig: DateConfig,
    restaurantId?: number
  ): Promise<ImportPreviewResponse> => {
    const response = await api.post('/imports/menus/preview', {
      file_id: fileId,
      column_mapping: columnMapping,
      date_config: dateConfig,
      restaurant_id: restaurantId,
    });
    return response.data;
  },

  confirm: async (request: ImportConfirmRequest): Promise<ImportConfirmResponse> => {
    const response = await api.post('/imports/menus/confirm', request);
    return response.data;
  },
};

export const catalogImportApi = {
  upload: async (file: File): Promise<CatalogImportUploadResponse> => {
    const formData = new FormData();
    formData.append('file', file);
    const response = await api.post('/imports/catalog/upload', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
      timeout: 30000,
    });
    return response.data;
  },

  preview: async (params: CatalogImportParams): Promise<CatalogImportPreviewResponse> => {
    const response = await api.post('/imports/catalog/preview', params);
    return response.data;
  },

  confirm: async (params: CatalogImportParams): Promise<CatalogImportResult> => {
    const response = await api.post('/imports/catalog/confirm', params);
    return response.data;
  },
};
